File: backend/app/api/schemas/statistics.py
"""Schemas for dashboard statistics endpoints."""

from datetime import date
from typing import Literal

from pydantic import BaseModel, Field


class CaptureDateCoverage(BaseModel):
    """How many of a project's media files have a capture date. Drives
    the "N images have no capture date" warning banner."""

    total: int
    without_date: int


class DashboardOverview(BaseModel):
    total_files: int
    total_observations: int
    total_events: int
    total_deployments: int
    total_sites: int
    trap_nights: int
    first_file_date: str | None
    last_file_date: str | None


class SpeciesCount(BaseModel):
    # `species` is the stable scientific-preferring key (used for colour and
    # selection). `common_name` is the parallel common label; the client
    # renders one or the other per the display preference.
    species: str
    common_name: str | None = None
    count: int
    # Distinct taxonomy IDs this row covers, for deep-linking the dashboard
    # bar to the Labels page filtered to these labels (F1). Empty for bars
    # with no taxonomy (e.g. "No taxonomy", or category-only rows).
    label_taxonomy_ids: list[str] = Field(default_factory=list)


class HourlyCount(BaseModel):
    hour: int
    count: int


class SunBands(BaseModel):
    """Fractional-hour timestamps for civil twilight + day boundaries.

    Computed server-side via python-astral from the project's averaged
    site lat/lon, its IANA timezone, and a reference date drawn from
    the filter range midpoint. Fed to the Activity pattern chart so
    the frontend can color each hour bar as night / dawn-dusk / day.
    """

    dawn: float
    sunrise: float
    sunset: float
    dusk: float


class ActivityPatternResponse(BaseModel):
    hours: list[HourlyCount]
    total_observations: int
    sun_bands: SunBands | None = None
    # Number of deployments in the filtered set that have no site
    # assigned. They are silently excluded from the sun-time
    # computation; the UI renders a banner when this is non-zero.
    deployments_without_site: int = 0


class DetectionTrendPoint(BaseModel):
    date: str
    count: int


class DetectionCategories(BaseModel):
    animal_count: int
    person_count: int
    vehicle_count: int
    empty_count: int


class LabelProgressRow(BaseModel):
    """Per-class verification progress for the dashboard's scrollable list.

    Counts are at the detection level: each detection has exactly one
    label, so rows partition cleanly. `scientific_name` is read from
    `label_taxonomy.scientific_name` when present, otherwise it falls back
    to the capitalised category for built-in (Person / Vehicle) and
    unclassified (Animal) detections.
    """

    label_taxonomy_id: str | None
    common_name: str
    scientific_name: str
    verified: int
    total: int


class VerificationProgressByLabel(BaseModel):
    rows: list[LabelProgressRow]


class SpeciesObservationCount(BaseModel):
    """One species (or category) and its MaxN sum within a single site.

    `common_name` / `scientific_name` are the two display strings; the
    client renders one per the display preference. `label` is the raw key.
    """

    label: str
    common_name: str
    scientific_name: str
    label_taxonomy_id: str | None
    count: int


class ObservationRateMapFeature(BaseModel):
    """One site plotted on the map.

    A site can have multiple deployments over time. `trap_nights` and
    `observation_count` are summed across the site's deployments;
    `rate_per_100` is the summed count divided by summed nights * 100,
    matching the dashboard's metric. `earliest_start_local` and
    `latest_end_local` describe the full monitoring range across the
    site's contributing deployments, not clipped to the active filter
    window. Sites without a `site_id` (deployments lacking a site
    assignment) are counted via `ObservationRateMapResponse.deployments_without_site`.
    """

    site_id: str
    site_name: str
    latitude: float
    longitude: float
    deployment_count: int
    earliest_start_local: date
    latest_end_local: date | None
    trap_nights: int
    observation_count: int
    rate_per_100: float
    species_breakdown: list[SpeciesObservationCount]


class ObservationRateMapResponse(BaseModel):
    features: list[ObservationRateMapFeature]
    # Number of deployments that otherwise passed the filters but had
    # no camera site assigned. Surfaced so the UI can render a
    # "X deployments without a site" banner.
    deployments_without_site: int = 0


# ---------------------------------------------------------------------------
# Activity overlap (Plots → Activity overlap page)
# ---------------------------------------------------------------------------


SampleSizeWarning = Literal["low_n_30", "low_n_50", "low_n_75"]
DielClass = Literal["diurnal", "nocturnal", "crepuscular", "cathemeral"]
DeltaEstimator = Literal["delta1", "delta4"]
TimeAxis = Literal["clock", "sun"]


class SpeciesActivity(BaseModel):
    """Per-species inputs to the activity-overlap chart.

    `kde_density` is the von Mises KDE evaluated on a 240-point grid
    over [0, 24) hours (matching `ml.activity_analysis.KDE_GRID_SAMPLES`),
    normalized so the integral over the grid is 1.0. The frontend draws
    this as a smooth curve.

    `raw_detection_times` is the underlying sample (decimal hours) used
    for the rug ticks under the curve. Capped at 5000 entries to bound
    payload size on huge datasets — the rug only needs visual density,
    not every single tick.

    `diel_density_by_phase` keys are "day", "night", "twilight" and sum
    to ~1.0. Surfaced so a UI tooltip can show the exact rule that
    produced the diel classification.
    """

    label: str
    n: int
    raw_detection_times: list[float]
    kde_density: list[float]
    diel_class: DielClass
    diel_density_by_phase: dict[str, float]
    sample_size_warning: SampleSizeWarning | None = None
    # Count of observations skipped because their date had no defined
    # sunrise (polar night / day). Always 0 in clock mode.
    dropped_polar: int = 0


class OverlapStat(BaseModel):
    """Pairwise activity overlap coefficient and bootstrap CI.

    Δ = ∫ min(f_a, f_b) dt over [0, 24], where f_a and f_b are the
    species' KDE densities. CI is from a 1000-rep percentile bootstrap.
    `delta_estimator` is the conventional Ridout & Linkie 2009 label
    (delta4 above min-N=50, delta1 below) — the underlying KDE method
    is the same for both, the label tells the reader which name to cite.
    """

    delta_estimator: DeltaEstimator
    delta: float
    ci_low: float
    ci_high: float
    bootstrap_reps: int
    min_n: int


class ActivityOverlapResponse(BaseModel):
    """Full payload for the Plots → Activity overlap page.

    `species_b` is None when the user has only picked one species; the
    chart still renders one curve in that case. `overlap` is None
    whenever we don't have two non-empty species to compare.

    `project_timezone` and `independence_interval_seconds` are
    read-only echoes of the two project settings that the user needs
    to know when reading the chart (x-axis interpretation and event
    grouping). The frontend surfaces them in the footer.
    """

    species_a: SpeciesActivity
    species_b: SpeciesActivity | None
    overlap: OverlapStat | None
    sun_bands: SunBands | None
    # Reference date the clock-mode `sun_bands` were computed for
    # (the midpoint of the filter range by default). Echoed so the
    # chart can caption the bands with the day they represent. Null
    # when sun_bands is null.
    sun_bands_reference_date: date | None = None
    # Mean dawn / sunrise / sunset / dusk across every observation's
    # date. Populated only in sun-time mode; the chart uses it to paint
    # twilight bands at the anchor positions.
    anchor_sun_bands: SunBands | None = None
    # Axis convention of the returned KDE: "clock" = raw wall-clock
    # hour, "sun" = Vazquez-anchored sun time.
    time_axis: TimeAxis = "clock"
    project_timezone: str
    independence_interval_seconds: int
    # Number of deployments in the filtered set that have no site
    # assigned. They are silently excluded from sun-time averaging;
    # the UI renders a banner when this is non-zero and sun mode is
    # active.
    deployments_without_site: int = 0

File: frontend/src/api/statistics.ts
/**
 * Statistics API endpoints.
 *
 * Following DEVELOPERS.md principles:
 * - Type hints everywhere
 * - Explicit operations
 */

import { api } from "../lib/api-client";

// --- Response types (matching backend schemas) ---

export interface DashboardOverview {
  total_files: number;
  total_observations: number;
  total_events: number;
  total_deployments: number;
  total_sites: number;
  trap_nights: number;
  first_file_date: string | null;
  last_file_date: string | null;
}

export interface CaptureDateCoverage {
  total: number;
  without_date: number;
}

export interface SpeciesCount {
  /** Stable scientific-preferring key (used for colour and selection). */
  species: string;
  common_name: string | null;
  count: number;
  /** Distinct taxonomy IDs this row covers, for deep-linking a dashboard
   *  bar to the Labels page filtered to these labels. Empty for bars with
   *  no taxonomy. */
  label_taxonomy_ids: string[];
}

export interface HourlyCount {
  hour: number;
  count: number;
}

export interface SunBands {
  /** Fractional hour (0-24) when civil twilight starts. */
  dawn: number;
  /** Fractional hour when the sun clears the horizon. */
  sunrise: number;
  /** Fractional hour when the sun drops below the horizon. */
  sunset: number;
  /** Fractional hour when civil twilight ends. */
  dusk: number;
}

export interface ActivityPatternResponse {
  hours: HourlyCount[];
  total_observations: number;
  /** Null when the project has no sites or astral can't compute. */
  sun_bands: SunBands | null;
  /** Deployments in the filtered set that have no camera site assigned.
   * They are silently excluded from sun-band averaging; the UI
   * surfaces a banner when this is non-zero. */
  deployments_without_site: number;
}

export interface DetectionTrendPoint {
  date: string;
  count: number;
}

export interface DetectionCategories {
  animal_count: number;
  person_count: number;
  vehicle_count: number;
  empty_count: number;
}

export interface LabelProgressRow {
  label_taxonomy_id: string | null;
  common_name: string;
  scientific_name: string;
  verified: number;
  total: number;
}

export interface VerificationProgressByLabel {
  rows: LabelProgressRow[];
}

export interface SpeciesObservationCount {
  label: string;
  common_name: string;
  scientific_name: string;
  label_taxonomy_id: string | null;
  count: number;
}

export interface ObservationRateMapFeature {
  site_id: string;
  site_name: string;
  latitude: number;
  longitude: number;
  /** Number of deployments that contribute to this site feature. */
  deployment_count: number;
  /** Min start_date across the site's contributing deployments. */
  earliest_start_local: string;
  /** Max end_date across the site's contributing deployments, or null
   * if no contributing deployment has an end date set (defensive). */
  latest_end_local: string | null;
  /** Summed trap nights across the site's deployments, clipped to the
   * active date filter window. */
  trap_nights: number;
  /** Summed MaxN observations across the site's deployments. */
  observation_count: number;
  rate_per_100: number;
  species_breakdown: SpeciesObservationCount[];
}

export interface ObservationRateMapResponse {
  features: ObservationRateMapFeature[];
  /** Deployments that passed the filter but have no site (and thus no
   * lat/lon). Surfaces in the map page as a banner. */
  deployments_without_site: number;
}

export interface ObservationRateMapFilters {
  siteIds?: string[];
  dateFrom?: string;
  dateTo?: string;
  labelTaxonomyIds?: string[];
}

// --- Activity overlap (Plots → Activity overlap page) ---

export type DielClass =
  | "diurnal"
  | "nocturnal"
  | "crepuscular"
  | "cathemeral";

export type DeltaEstimator = "delta1" | "delta4";

export type SampleSizeWarning = "low_n_30" | "low_n_50" | "low_n_75";

export type TimeAxis = "clock" | "sun";

export interface SpeciesActivity {
  label: string;
  n: number;
  /** Decimal hours [0..24) for the rug ticks under the curve. Capped at 5000. */
  raw_detection_times: number[];
  /** 240-point von Mises KDE over [0..24), normalized to integrate to 1. */
  kde_density: number[];
  diel_class: DielClass;
  /** {"day": number, "night": number, "twilight": number} summing to ~1. */
  diel_density_by_phase: Record<string, number>;
  sample_size_warning: SampleSizeWarning | null;
  /** Observations skipped because their date had no defined sunrise (polar). */
  dropped_polar: number;
}

export interface OverlapStat {
  delta_estimator: DeltaEstimator;
  delta: number;
  ci_low: number;
  ci_high: number;
  bootstrap_reps: number;
  min_n: number;
}

export interface ActivityOverlapResponse {
  species_a: SpeciesActivity;
  species_b: SpeciesActivity | null;
  overlap: OverlapStat | null;
  /** Single-reference clock-time bands (midpoint date, project avg lat/lon). */
  sun_bands: SunBands | null;
  /** ISO date (YYYY-MM-DD) the clock-mode bands were computed for. Null when sun_bands is null. */
  sun_bands_reference_date: string | null;
  /** Mean-anchored dawn/sunrise/sunset/dusk across observations. Non-null only in sun mode. */
  anchor_sun_bands: SunBands | null;
  /** The axis the returned KDE is in. May differ from the requested axis when
   * sun mode was not possible (no site coordinates, all-polar dataset) and
   * the backend silently downgraded to clock. */
  time_axis: TimeAxis;
  /** IANA timezone the project's camera clocks are set to. */
  project_timezone: string;
  independence_interval_seconds: number;
  /** Deployments in the filtered set that have no camera site assigned.
   * They are silently excluded from sun-band averaging; the UI
   * surfaces a banner in sun mode when this is non-zero. */
  deployments_without_site: number;
}

export interface ActivityOverlapFilters {
  speciesA: string;
  speciesB?: string;
  siteIds?: string[];
  dateFrom?: string;
  dateTo?: string;
  taxonomicRank?: string;
  timeAxis?: TimeAxis;
}

// --- Shared helpers ---

/**
 * Build query string with project_id and optional filter params.
 * Only includes params that are actually provided.
 */
function buildParams(
  projectId: string,
  options?: { species?: string; siteIds?: string; dateFrom?: string; dateTo?: string; taxonomicRank?: string }
): string {
  const params = new URLSearchParams();
  params.set("project_id", projectId);

  if (options?.species) params.set("species", options.species);
  if (options?.siteIds) params.set("site_ids", options.siteIds);
  if (options?.dateFrom) params.set("date_from", options.dateFrom);
  if (options?.dateTo) params.set("date_to", options.dateTo);
  if (options?.taxonomicRank) params.set("taxonomic_rank", options.taxonomicRank);

  return params.toString();
}

// --- API client ---

export const statisticsApi = {
  /**
   * Dashboard overview counts (files, detections, events, etc.)
   */
  getOverview: (projectId: string, siteIds?: string, dateFrom?: string, dateTo?: string) => {
    const query = buildParams(projectId, { siteIds, dateFrom, dateTo });
    return api.get<DashboardOverview>(`/api/statistics/overview?${query}`);
  },

  /** Media-file count + how many have no capture date (warning banner). */
  getCaptureDateCoverage: (projectId: string) => {
    const query = buildParams(projectId, {});
    return api.get<CaptureDateCoverage>(
      `/api/statistics/capture-date-coverage?${query}`,
    );
  },

  /**
   * Species distribution (species name + detection count)
   */
  getSpeciesDistribution: (projectId: string, siteIds?: string, dateFrom?: string, dateTo?: string, taxonomicRank?: string, countMode?: string, wildlifeOnly?: boolean) => {
    const query = buildParams(projectId, { siteIds, dateFrom, dateTo, taxonomicRank });
    const modeParam = countMode ? `&count_mode=${countMode}` : "";
    // wildlife_only drops person/vehicle categories and non-wildlife
    // labels (blank, bait, human, ...); the dashboard bars use it, the
    // species pickers keep the full list.
    const wildlifeParam = wildlifeOnly ? "&wildlife_only=true" : "";
    // Returns every matching species; the dashboard bars slice to 10 client-side.
    return api.get<SpeciesCount[]>(`/api/statistics/species?${query}${modeParam}${wildlifeParam}`);
  },

  /**
   * Hourly activity pattern, optionally filtered by species
   */
  getActivityPattern: (
    projectId: string,
    params?: { species?: string; siteIds?: string; dateFrom?: string; dateTo?: string; taxonomicRank?: string }
  ) => {
    const query = buildParams(projectId, params);
    return api.get<ActivityPatternResponse>(`/api/statistics/activity-pattern?${query}`);
  },

  /**
   * Daily detection trend over time, optionally filtered by species
   */
  getDetectionTrend: (
    projectId: string,
    params?: { species?: string; siteIds?: string; dateFrom?: string; dateTo?: string; taxonomicRank?: string }
  ) => {
    const query = buildParams(projectId, params);
    return api.get<DetectionTrendPoint[]>(`/api/statistics/detection-trend?${query}`);
  },

  /**
   * Detection category breakdown (animal, person, vehicle, empty)
   */
  getDetectionCategories: (projectId: string, siteIds?: string, dateFrom?: string, dateTo?: string) => {
    const query = buildParams(projectId, { siteIds, dateFrom, dateTo });
    return api.get<DetectionCategories>(`/api/statistics/categories?${query}`);
  },

  /**
  /**
   * Per-class verification progress. Each row is one label; counts are
   * verified / total detections that pass the project floor (with the
   * verified-override). Sorted by total descending. Excludes the
   * "false detection" label.
   */
  getVerificationProgressByLabel: (projectId: string, siteIds?: string, dateFrom?: string, dateTo?: string) => {
    const query = buildParams(projectId, { siteIds, dateFrom, dateTo });
    return api.get<VerificationProgressByLabel>(
      `/api/statistics/verification-progress-by-label?${query}`,
    );
  },

  /**
   * Per-site observation rate features for the Map page. Each feature
   * aggregates across the site's contributing deployments.
   * Rate = observations / trap nights * 100, where observations is
   * sum(EventObservation.max_n) per event.
   */
  getObservationRateMap: (
    projectId: string,
    filters?: ObservationRateMapFilters
  ) => {
    const params = new URLSearchParams();
    params.set("project_id", projectId);
    if (filters?.siteIds?.length) {
      params.set("site_ids", filters.siteIds.join(","));
    }
    if (filters?.dateFrom) params.set("date_from", filters.dateFrom);
    if (filters?.dateTo) params.set("date_to", filters.dateTo);
    if (filters?.labelTaxonomyIds?.length) {
      params.set("label_taxonomy_ids", filters.labelTaxonomyIds.join(","));
    }
    return api.get<ObservationRateMapResponse>(
      `/api/statistics/observation-rate-map?${params.toString()}`
    );
  },

  /**
   * Activity overlap payload for the Plots → Activity overlap page.
   * Returns 1- or 2-species KDE curves, sun bands, diel classification,
   * and the Ridout & Linkie overlap coefficient Δ with bootstrap CI
   * (when both species have data).
   */
  getActivityOverlap: (
    projectId: string,
    filters: ActivityOverlapFilters
  ) => {
    const params = new URLSearchParams();
    params.set("project_id", projectId);
    params.set("species_a", filters.speciesA);
    if (filters.speciesB) params.set("species_b", filters.speciesB);
    if (filters.siteIds?.length) {
      params.set("site_ids", filters.siteIds.join(","));
    }
    if (filters.dateFrom) params.set("date_from", filters.dateFrom);
    if (filters.dateTo) params.set("date_to", filters.dateTo);
    if (filters.taxonomicRank) params.set("taxonomic_rank", filters.taxonomicRank);
    if (filters.timeAxis) params.set("time_axis", filters.timeAxis);
    return api.get<ActivityOverlapResponse>(
      `/api/statistics/activity-overlap?${params.toString()}`
    );
  },
};
